graphix/open_graph.py

```diff
@@ -0,0 +1,248 @@
+    """
+
+    angle: float
+    plane: Plane
+
+    def __eq__(self, other: object) -> bool:
+        """Compares if two measurements are equal
+
+        Example
+        -------
+        >>> from graphix.open_graph import Measurement
+        >>> from graphix.pauli import Plane
+        >>> Measurement(0.0, Plane.XY) == Measurement(0.0, Plane.XY)
+        True
+        >>> Measurement(0.0, Plane.XY) == Measurement(0.0, Plane.YZ)
+        False
+        >>> Measurement(0.1, Plane.XY) == Measurement(0.0, Plane.XY)
+        False
+        """
+        if not isinstance(other, Measurement):
+            return NotImplemented
+
+        return np.allclose(self.angle, other.angle) and self.plane == other.plane
+
+    def is_z_measurement(self) -> bool:
+        """Indicates whether it is a Z measurement
+
+        Example
+        -------
+        >>> from graphix.open_graph import Measurement
+        >>> Measurement(0.0, Plane.XY).is_z_measurement()
+        True
+        >>> Measurement(0.0, Plane.YZ).is_z_measurement()
+        False
+        >>> Measurement(0.1, Plane.XY).is_z_measurement()
+        False
+        """
+        return np.allclose(self.angle, 0.0) and self.plane == Plane.XY
+
+
+class OpenGraph:
+    """Open graph contains the graph, measurement, and input and output
+    nodes. This is the graph we wish to implement deterministically
+
+    :param inside: the underlying graph state
+    :param measurements: a dictionary whose key is the ID of a node and the
+        value is the measurement at that node
+    :param inputs: a set of IDs of the nodes that are inputs to the graph
+    :param outputs: a set of IDs of the nodes that are outputs of the graph
+
+    Example
+    -------
+    >>> import networkx as nx
+    >>> from graphix.open_graph import OpenGraph, Measurement
+    >>>
+    >>> inside_graph = nx.Graph([(0, 1), (1, 2), (2, 0)])
+    >>>
+    >>> measurements = {i: Measurement(0.5 * i, Plane.XY) for i in range(2)}
+    >>> inputs = [0]
+    >>> outputs = [2]
+    >>> og = OpenGraph(inside_graph, measurements, inputs, outputs)
+    """
+
+    # All the measurement, inputs and outputs information is stored in this
+    # graph. This is to simplify certain operations such as composition of
+    # graphs and checking equality.
+    # These attributes shouldn't be accessed directly by the user, instead,
+    # they should use the inputs(), outputs(), and measurements() properties.
+    #
+    # The following attributes are stored in the nodes:
+    #
+    # Non-output nodes:
+    # * "measurement" - The Measurement object for the node
+    #
+    # Input nodes:
+    # * "is_input"    - The value is always True
+    # * "input_order" - A zero-indexed integer used to preserve the ordering of
+    #                   the inputs
+    #
+    # Output nodes:
+    # * "is_output"    - The value is always True
+    # * "output_order" - A zero-indexed integer used to preserve the ordering of
+    #                    the outputs
```

Comment:
Why do we care about preserving the order of inputs and outputs but not the order of measurements? In the same vein, this representation forgets s-domains and t-domains: I understand that measurement orders and domains are irrelevant if the computation is deterministic but we should take care of not restricting ourselves to this case, at least when defining data structures (whereas it is fine for some algorithms to be specific to a certain case of flow for instance, such data structures should be general). I would prefer if open graphs were isomorphic to standardized patterns, that is if we preserve input orders, output orders, measurement orders and byproducts.

Reply:
Hi @thierry-martinez ! I have a little different perspective on your comment.

From the relationship between ZX Calculus and MBQC, isn’t it more natural that open graph class is equivalent to ZX diagram, especially, MBQC-LC form, not to the pattern? If we want to design universal non-Unitary pattern, I think one of the natural ways is to include freedom of correction set design into the open graph, which can be done in a separated module. In fact, some of papers studying non-Unitary computation on MBQC do not use command sequence in their study because pattern is too much information for their purpose. It may degrade the abstractness of open graph. What do you think about it?

Reply:
@thierry-martinez I agree with @masa10-f. the exact form of correction sets are independent from open graph feature to be implemented here (main purpose is to have a common interface between patterns and zx diagrams), and is something to be dealt with in a separate PR.